include/deal.II/fe/fe_rt_bubbles.h

```diff
@@ -0,0 +1,146 @@
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii__fe_raviart_thomas_bubbles_h
+#define dealii__fe_raviart_thomas_bubbles_h
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/table.h>
+#include <deal.II/base/polynomials_rt_bubbles.h>
+#include <deal.II/base/polynomial.h>
+#include <deal.II/base/tensor_product_polynomials.h>
+#include <deal.II/base/geometry_info.h>
+#include <deal.II/fe/fe.h>
+#include <deal.II/fe/fe_poly_tensor.h>
+
+#include <vector>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * This class implements a curl-enhanced Raviart-Thomas elements,
+ * conforming with <i>H<sup>div</sup></i> space. The node functionals are
+ * defined as point values in Gauss-Lobatto points. These elements generate
+ * vector fields with normal components continuous between mesh cells. The
+ * purpose of this finite element is in localizing the interactions
+ * between degrees of freedom around the nodes when an appropriate quadrature
+ * rule is used, leading to a block-diagonal mass matrix (even with full-tensor
+ * coefficient).
+ *
+ * The elements are defined through enrichment of classical Raviart-Thomas
+ * elements with extra curls, so that the <i>H<sup>div</sup></i> conformity is
+ * preserved, and the total number of degrees of freedom of FE_RT_Bubbles of
+ * order k is equal to the number of DoFs in <i>dim</i> copies of FE_Q of
+ * order <i>k</i>.
+ *
+ * @note Unlike Raviart-Thomas, the lowest possible order for this
+ * enhanced finite element is 1, i.e. $k \ge 1$.
+ *
+ * The matching pressure space for FE_RT_Bubbles of order <i>k</i> is FE_DGQ of
+ * order <i>k-1</i>. With the exact integration, this pair yields $(k+1)$-st
+ * order of convergence in $L_2$-norm for a vector variable and $k$-th order in
+ * $L_2$-norm for a scalar one (same as $BDM_k \times P_{k-1}$).
+ *
+ * For this enhanced Raviart-Thomas element, the node values are not cell
+ * and face moments with respect to certain polynomials, but the values in
+ * Gauss-Lobatto quadrature points. The nodal values on edges (faces in <i>3D</i>)
+ * are evaluated first, according to the natural ordering of the edges (faces) of
+ * a cell. The interior degrees of freedom are evaluated last.
+ *
+ * For an RT-Bubbles element of degree <i>k</i>, we choose
+ * <i>(k+1)<sup>dim-1</sup></i> Gauss-Lobatto points on each face. These points
+ * are ordered lexicographically with respect to the orientation of the face.
+ * In the interior of the cells, the values are computed using an anisotropic
+ * Gauss-Lobatto formula for integration. The mass matrix assembled with the
+ * use of this same quadrature rule, is block diagonal with blocks
+ * corresponding to quadrature points. See
+ * <i><a href="https://arxiv.org/abs/1710.06742">"Higher order multipoint flux mixed
+ * finite element methods on quadrilaterals and hexahedra"</a></i> for more details.
```

Comment:
Great, thanks for adding this.

Do the results shown there use deal.II? If so, we would appreciate if you could add one of the references from http://dealii.org/publications.html to the paper before publication.

Reply:
Yes, deal.II is referenced in the newer version of the paper, the arXiv version will be updated by Monday 15th. This reference will be included in published version as well.